### ISSUE TITLE
Duplicate Message in Realtime Tables with Kafka

### ISSUE THREAD

**Issue by ankitsultana** [apache/pinot#10159]
We have been running the latest master on one of our clusters and one of our customers reported seeing duplicate messages in their realtime tables. The issue continues to happen even after we have switched to using read_committed isolation level.

Filing this issue for tracking this. Will add more details soon.

cc: @Jackie-Jiang

**Comment by Jackie-Jiang** [comment]
@navina Can you help take a look at this issue?

**Comment by navina** [comment]
@ankitsultana can you provide more info about this? Are you able to reproduce this in your local dev envt?

**Comment by ankitsultana** [comment]
Thanks Jackie and Navina. One of the engineers on our end is looking into this as well and it might be due to producer side code.

Let me get back to this by early next week.

**Comment by ankitsultana** [comment]
Based on some analysis it looks like the issue was happening at the Kafka level. Apologies for the confusion here.

**Comment by navina** [comment]
@ankitsultana  no worries. can you provide more details on the issue happening at Kafka level?

**Comment by ankitsultana** [comment]
We didn't have exactly once guarantees on both producer and consumer side. In our first investigation we had found duplicates only in Pinot but not in Kafka so we suspected it could be a bug in the consumer side. But on taking a deeper look we found both cases: duplicates in Kafka and duplicates in Pinot, and duplicates only in Pinot.

This particular use-case had moderately high message throughput. Our suspicion is that it may be more likely to hit issues like duplicates.
